rfcs/20230704-dynamism-101.md

```diff
@@ -0,0 +1,825 @@
+(StableHLO project can only provide compatibility guarantees for the StableHLO
+opset).
+
+<a id="f6" /> **(F6)** Dimension numbers and dimension sizes are modelled
+inconsistently. For example, PadOp represents paddings as arrays of `i64`.
+However, DynamicPadOp takes all sorts of paddings - tensors of `index`,
+tensors of `i32`, tensors of `i64`, etc.
+
+<a id="f7" /> **(F7)** There is alignment on shape mismatches being undefined
+behavior, but there are multiple schools of thought on how this undefined
+behavior should be guarded against (use `shape` dialect, use asserts, don't do
+anything).
+
+<a id="f8" /> **(F8)**. Even though StableHLO dynamism is used fairly actively,
+it is not reflected in the StableHLO specification. As a result, dynamism
+semantics and especially constraints are underspecified.
+
+## Out of scope
+
+This RFC aims to address a considerable part of community feedback on the
+existing design of dynamism in MHLO and StableHLO, but some feedback is
+deliberately out of scope because further discussion is needed to obtain
+alignment in the corresponding areas.
+
+<a id="o1" /> **(O1)** There is promising related work on alternative
+representations for dynamism, e.g. involving symbols and formulas in modeling
+dynamic sizes rather than representing dynamic sizes as question marks. These
+representations are more expressive than the design proposed in this RFC, and
+they have the potential to solve the problem of shape mismatches, to convey
+additional information to consumers improving overall performance, etc.
+
+However, this is a very new area for the OpenXLA stack, so a considerable
+amount of design exploration will be needed before these ideas can be turned
+into concrete proposals. Meanwhile, this RFC is focused on making design
+improvements which have been already socialized within the community.
+
+<a id="o2" /> **(O2)** Shape computations are considerably improved by this RFC,
+but not all of the feedback from [F5](#f5)/[F6](#f6) is addressed.
+
+More specifically, this RFC proposes to represent shape computations as
+StableHLO operations working on 0-dimensional tensors, which is a practically
+important simplification for StableHLO programs and their producers/consumers.
+However, interoperability with other dialects, including `arith` and `shape`
+which are oftentimes used by the community for shape computations, is out of
+scope, and [P6](#p6) goes into details about the rationale.
+
+<a id="o3" /> **(O3)** Working out a common solution for shape checks would be
+nice. Per [F7](#f7), there are multiple incompatible approaches, so a common
+solution would result in a simplification of the overall stack.
+
+However, based on author's experience, alignment in this area requires
+non-trivial effort (different approaches have different benefits, plus they have
+limited interoperability), so this is left for future work.
+
+<a id="o4" /> **(O4)** Shape inference is an essential part of the StableHLO
```

Comment:
Does that mean that Shape Inference is provided by StableHLO? As in front ends shouldn't do their own?

Reply:
Shape inference is implemented by the StableHLO dialect for the majority of the StableHLO ops (there are some ops where shape inference is infeasible, and some ops which have known issues). However, even with shape inference provided by libStablehlo, sometimes producers may want to specify their own result types. #690 is one example pertaining to dynamism, but there are other dynamism-, quantization- and sparsity-related examples as well. Overall, this is complicated topic, which is why I didn't include it in this RFC :)

Reply:
Shape inference in StableHLO is part of the builder interface. When a programmer writes `builder.create<stablehlo::FooOp>(...)`, for most of the ops they can skip providing the result type and it will be inferred. For static shapes, shape inference works very well, so it is idiomatic to use it. For dynamic shapes, sometimes it's overly conservative, so the programmer may decide to provide the result type explicitly.

This is just a summary of what's going on, but there's more. Let's follow up on that offline since shape inference is out of scope for this RFC.